hieroplatform/src/main/java/org/hiero/dataset/RemoteDataSet.java

```diff
@@ -18,87 +18,208 @@
+     */
     @Override
-    public <S> Observable<PartialResult<IDataSet<Pair<T, S>>>> zip(
-            final IDataSet<S> other) {
+    public <S> Observable<PartialResult<IDataSet<Pair<T, S>>>> zip(final IDataSet<S> other) {
         if (!(other instanceof RemoteDataSet<?>)) {
             throw new RuntimeException("Unexpected type in Zip " + other);
         }
         final RemoteDataSet<S> rds = (RemoteDataSet<S>) other;
 
         // zip commands are not valid if the RemoteDataSet instances point to different
         // actor systems or different nodes.
-        final Address leftAddress = this.remoteActor.path().address();
-        final Address rightAddress = rds.remoteActor.path().address();
+        final HostAndPort leftAddress = this.serverEndpoint;
+        final HostAndPort rightAddress = rds.serverEndpoint;
         if (!leftAddress.equals(rightAddress)) {
-            throw new RuntimeException("Zip command invalid for RemoteDataSets across different servers" +
-                    "| left: " + leftAddress + ", right:" + rightAddress);
+            throw new RuntimeException("Zip command invalid for RemoteDataSets " +
+                    "across different servers | left: " + leftAddress + ", right:" + rightAddress);
+        }
+
+        final ZipOperation zip = new ZipOperation(rds.remoteHandle);
+        final byte[] serializedOp = SerializationUtils.serialize(zip);
+        final Command command = Command.newBuilder()
+                                         .setIdsIndex(this.remoteHandle)
+                                         .setSerializedOp(ByteString.copyFrom(serializedOp))
+                                         .build();
+        final PublishSubject<PartialResult<IDataSet<Pair<T, S>>>> subj = PublishSubject.create();
+        final StreamObserver<PartialResponse> responseObserver =
+                                                        new NewDataSetObserver<Pair<T, S>>(subj);
+        return subj.doOnSubscribe(() -> this.stub.withDeadlineAfter(TIMEOUT, TimeUnit.MILLISECONDS)
+                                                 .zip(command, responseObserver))
+                   .doOnUnsubscribe(() -> unsubscribe(zip.id));
+    }
+
+    /**
+     * Unsubscribes an operation. This method is safe to invoke multiple times because the
+     * logic on the remote end is idempotent.
+     */
+    private void unsubscribe(UUID id) {
+        UnsubscribeOperation op = new UnsubscribeOperation(id);
+        final byte[] serializedOp = SerializationUtils.serialize(op);
+        final Command command = Command.newBuilder()
+                                       .setIdsIndex(this.remoteHandle)
+                                       .setSerializedOp(ByteString.copyFrom(serializedOp))
+                                       .build();
+        this.stub.withDeadlineAfter(TIMEOUT, TimeUnit.MILLISECONDS)
+                 .unsubscribe(command, new StreamObserver<Ack>() {
+            @Override
```

Comment:
I'm not sure myself. An upstream caller would need to subscribe to an unsubscribe event with a doOnUnsubscribe, which it will eventually be notified about. I should write a test for this.